vyper/venom/__init__.py

```diff
@@ -36,102 +34,272 @@
+    OptimizationLevel.Os
+]  # codesize -> Os
+
 
 def generate_assembly_experimental(
     venom_ctx: IRContext, optimize: OptimizationLevel = DEFAULT_OPT_LEVEL
 ) -> list[AssemblyInstruction]:
     compiler = VenomCompiler(venom_ctx)
-    return compiler.generate_evm_assembly(optimize == OptimizationLevel.NONE)
-
-
-def _run_passes(fn: IRFunction, optimize: OptimizationLevel, ac: IRAnalysesCache) -> None:
-    # Run passes on Venom IR
-    # TODO: Add support for optimization levels
-
-    FloatAllocas(ac, fn).run_pass()
-
-    SimplifyCFGPass(ac, fn).run_pass()
-
-    MakeSSA(ac, fn).run_pass()
-    PhiEliminationPass(ac, fn).run_pass()
-
-    # run constant folding before mem2var to reduce some pointer arithmetic
-    AlgebraicOptimizationPass(ac, fn).run_pass()
-    SCCP(ac, fn, remove_allocas=False).run_pass()
-    SimplifyCFGPass(ac, fn).run_pass()
-
-    AssignElimination(ac, fn).run_pass()
-    Mem2Var(ac, fn).run_pass()
-    MakeSSA(ac, fn).run_pass()
-    PhiEliminationPass(ac, fn).run_pass()
-    SCCP(ac, fn).run_pass()
-
-    SimplifyCFGPass(ac, fn).run_pass()
-    AssignElimination(ac, fn).run_pass()
-    AlgebraicOptimizationPass(ac, fn).run_pass()
-
-    LoadElimination(ac, fn).run_pass()
-    PhiEliminationPass(ac, fn).run_pass()
-    AssignElimination(ac, fn).run_pass()
-
-    SCCP(ac, fn).run_pass()
-    AssignElimination(ac, fn).run_pass()
-    RevertToAssert(ac, fn).run_pass()
-
-    SimplifyCFGPass(ac, fn).run_pass()
-    MemMergePass(ac, fn).run_pass()
-    RemoveUnusedVariablesPass(ac, fn).run_pass()
-
-    DeadStoreElimination(ac, fn).run_pass(addr_space=MEMORY)
-    DeadStoreElimination(ac, fn).run_pass(addr_space=STORAGE)
-    DeadStoreElimination(ac, fn).run_pass(addr_space=TRANSIENT)
-    LowerDloadPass(ac, fn).run_pass()
+    return compiler.generate_evm_assembly(
+        optimize in (OptimizationLevel.NONE, OptimizationLevel.O0)
+    )
 
-    BranchOptimizationPass(ac, fn).run_pass()
 
-    AlgebraicOptimizationPass(ac, fn).run_pass()
+# Mapping of pass names to their disable flag names
+# Passes not in this map are considered essential and always run
+PASS_FLAG_MAP = {
+    "AlgebraicOptimizationPass": "disable_algebraic_optimization",
+    "SCCP": "disable_sccp",
+    "Mem2Var": "disable_mem2var",
+    "LoadElimination": "disable_load_elimination",
+    "RemoveUnusedVariablesPass": "disable_remove_unused_variables",
+    "DeadStoreElimination": "disable_dead_store_elimination",
+    "BranchOptimizationPass": "disable_branch_optimization",
+    "CSE": "disable_cse",
+    "SimplifyCFGPass": "disable_simplify_cfg",
+}
 
-    # This improves the performance of cse
-    RemoveUnusedVariablesPass(ac, fn).run_pass()
 
-    PhiEliminationPass(ac, fn).run_pass()
-    AssignElimination(ac, fn).run_pass()
-    CSE(ac, fn).run_pass()
+def _run_passes(fn: IRFunction, flags: VenomOptimizationFlags, ac: IRAnalysesCache) -> None:
+    passes = OPTIMIZATION_PASSES.get(flags.level, OPTIMIZATION_PASSES[OptimizationLevel.O2])
```

Comment:
probably better to fail with a key lookup error since it indicates a mistake in the optimization passes dict (or how we have constructed flags.level)